src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,407 @@
+[4] https://github.com/Z3Prover/z3/blob/d047b86439ec209446d211f0f6b251ebfba070d8/src/ackermannization/lackr_model_constructor.cpp#L344
+-/
+prelude
+import Lean.Expr
+import Lean.Message
+import Std.Tactic.BVDecide.Bitblast
+import Std.Tactic.BVAckermannize.Syntax
+import Lean.Meta.Basic
+import Lean.Elab.Tactic.Basic
+import Lean.Meta.LitValues
+import Lean.Meta.InferType
+import Lean.Elab.Tactic.FalseOrByContra
+import Lean.Meta.Tactic.Assert
+import Lean.Meta.Tactic.Util
+import Lean.Meta.Tactic.Congr
+import Lean.Meta.Tactic.Replace
+
+open Lean Elab Meta Tactic
+
+initialize Lean.registerTraceClass `bv_ack
+
+namespace Ack
+
+ /-- Types that can be bitblasted by bv_decide -/
+ inductive BVTy
+ /-- Booleans -/
+ | Bool
+ /-- Bitvectors of a fixed width `w` -/
+ | BitVec (w : Nat)
+ deriving Hashable, DecidableEq, Inhabited
+
+instance : ToMessageData BVTy where
+  toMessageData
+  | .Bool => m!"bool"
+  | .BitVec w => m!"BitVec {w}"
+
+namespace BVTy
+
+/-- Reify a raw expression of type `Type` into the types of bitvectors we can bitblast,
+returning `none` if `e` was not recognized as either `Bool` or `BitVec ?w`,
+with `?w` a literal `Nat`  -/
+def ofExpr? (e : Expr) : OptionT MetaM BVTy :=
+  match_expr e.consumeMData with
+  | _root_.Bool => return Bool
+  | _root_.BitVec w => do
+     let w ← getNatValue? w
+     return .BitVec w
+  | _ => OptionT.fail
+
+/-- Convert a `BVTy` back into an `Expr` -/
+def toExpr : BVTy → Expr
+| .Bool => mkConst ``_root_.Bool
+| .BitVec w => mkApp (mkConst ``_root_.BitVec) (mkNatLit w)
+
+end BVTy
+
+/-- An argument to an uninterpreted function, which we track for ackermannization. -/
+structure Argument where
+  /-- The expression corresponding to the argument -/
+  x : Expr
+  /-- The cached type of the expression x -/
+  xTy : BVTy
+deriving Hashable, BEq, Inhabited
+
+namespace Argument
+
+instance : ToMessageData Argument where
+  toMessageData arg := m!"{arg.x} : {arg.xTy}"
+
+/-- Build an `Argument` from a raw expression of type `Bool` or `BitVec _`. -/
+def ofExpr? (e : Expr) : OptionT MetaM Argument := do
+  let t ← BVTy.ofExpr? (← inferType e)
+  return { x := e, xTy := t}
+
+end Argument
+
+/--
+A function, which packs the expression and the type of the codomain of the function.
+We use the type of the codomain to build an abstracted value for every call of this function.
+-/
+structure Function where
+  /-- The function -/
+  f : Expr
+  /-- The type of the function's codomain -/
+  codomain : BVTy
+ deriving Hashable, BEq, Inhabited
+namespace Function
+
+instance : ToMessageData Function where
+  toMessageData fn := m!"{fn.f} (cod: {fn.codomain})"
+
+/--
+Reify an expression `e` of the kind `f x₁ ... xₙ`, where all the arguments and the return type are
+a `BVTy` into an App.
+-/
+def reifyAp (f : Expr) : OptionT MetaM (Function × Array Argument) := do
+  let xs := f.getAppArgs
```

Comment:
what if we are working on a function like
```
f : Nat -> Bool -> Bool
```
applied to something like `f 1 b`. In this situation you would fully give up on reification even though it is very much possible to do this by using `(f 1)` as the function instead. Do we want this? And if so, how far do we want to push it? In theory we could even support things like `f : Nat -> Bool -> Nat -> Bool`.

Reply:
Yes, that's a fair point, I do indeed bail out. I added a comment describing that it's currently very conservative, and that we can extend to handle a larger fragment.